src/dhtproto/client/legacy/internal/helper/Handshake.d

```diff
@@ -0,0 +1,241 @@
+            been reached
+
+        ***********************************************************************/
+
+        private void allNodesConnected ()
+        {
+            this.outer.log.info("All DHT nodes connected!");
+            this.all_connected_event.trigger();
+        }
+
+        /***********************************************************************
+
+            Runs the task
+
+        ***********************************************************************/
+
+        protected override void run ()
+        {
+            this.all_connected_event.wait();
+        }
+    }
+
+
+    /***************************************************************************
+
+        Task that suspends until at least one node is connected
+
+    ***************************************************************************/
+
+    private class OneDhtNodeConnected : Task
+    {
+        import swarm.Const : NodeItem;
+
+        /// Used to manage suspend/resume of this task
+        private TaskEvent node_connected_event;
+
+        /***********************************************************************
+
+            Delegate to be called by the handshake each time an individual
+            node connects
+
+            Params:
+                item = data on the address and port of the connected node
+
+        ***********************************************************************/
+
+        private void oneNodeConnected (NodeItem item)
+        {
+            this.outer.log.trace("Connected to DHT node {}:{}",
+                                 item.Address, item.Port);
+            this.node_connected_event.trigger();
+        }
+
+        /***********************************************************************
+
+            Runs the task
+
+        ***********************************************************************/
+
+        protected override void run ()
+        {
+            this.node_connected_event.wait();
+        }
+    }
+
+
+    /// Logger for the handshake
+    private Logger log;
+
+
+    /// Instance of task that suspends until all nodes are connected
+    private AllDhtNodesConnected all_nodes_connected;
+
+
+    /// Instance of task that suspends until at least one node is connected
+    private OneDhtNodeConnected one_node_connected;
+
+
+    /// Instance of class that launches the actual handshake, retrying when
+    /// it fails until all nodes are connected
+    private RetryHandshake retry_handshake;
+
+
+    /***************************************************************************
+
+        Constructor
+
+        Params:
+            dht = DHT client instance via which to perform the handshake
+            wait_time = time to wait (in seconds) before retrying failed
+                handshake attempts
+
+    ***************************************************************************/
+
+    public this (DhtClient dht, ulong wait_time)
```

Comment:
Nitpick: I like `retry_delay_seconds` (as used in the example) more than `wait_time`.

Reply:
Sure, can do.  I was copying `wait_time` from `RetryHandshake`, but I agree `retry_delay_seconds` is nicer.